rfcs/20231012-external-mel-bindings.md

```diff
@@ -0,0 +1,546 @@
+let root = dirname "/User/github"
+```
+
+If the name of the field imported starts with an invalid character in OCaml,
+like uppercase characters, `mel.as` can be used:
+
+```ocaml
+let%js.import { foo : string [@mel.as "Foo"] } = "./bar"
+```
+
+## `mel.get` and `mel.set`
+
+Current:
+
+```ocaml
+external set_title : document -> string -> unit = "title" [@@mel.set]
+external get_title : document -> string = "title" [@@mel.get]
+```
+
+Proposed:
+
+```ocaml
+let%js set_title : doc:document -> title:string -> unit = "$doc.title = $title"
+let%js get_title : doc:document -> string = "$doc.title"
+```
+
+## `mel.get_index` and `mel.set_index`
+
+Current:
+
+```ocaml
+type t
+external create : int -> t = "Int32Array" [@@mel.new]
+external get : t -> int -> int = "" [@@mel.get_index]
+external set : t -> int -> int -> unit = "" [@@mel.set_index]
+```
+
+Proposed:
+
+```ocaml
+type t
+let%js create : length:int -> t = "new Int32Array($length)"
+let%js get : arr:t -> pos:int -> int = "$arr[$pos]"
+let%js set : arr:t -> pos:int -> value:int -> unit = "$arr[$pos] = $value"
+```
+
+Let's see another example, in combination with importing modules:
+
+Before:
+
+```ocaml
+type t
+external book : unit -> t = "Book" [@@mel.new] [@@mel.module]
+let myBook = book ()
+```
+
+After:
+
+```ocaml
+type t
+let%js.import book: t = "Book"
+let%js create_book : ~t_:t -> unit -> t = "new $t_()"
+let myBook = book |> create_book ()
+```
+
+## `mel.send` and `mel.send.pipe`
+
+Current:
+
+```ocaml
+external get_by_id : document -> string -> Dom.element = "getElementById"
+  [@@mel.send]
+external get_by_id : string -> Dom.element = "getElementById"
+  [@@mel.send.pipe: document]
+```
+
+Proposed:
+
+```ocaml
+let%js get_by_id : doc_:document -> id:string -> Dom.element = "$doc_.getElementById($id)"
+let%js get_by_id : id:string -> doc_:document -> Dom.element = "$doc_.getElementById($id)"
+```
+
+## `mel.new`
+
+Current:
+
+```ocaml
+type t
+external create_date : unit -> t = "Date" [@@mel.new]
+let date = create_date ()
+```
+
+Proposed:
+
+```ocaml
+type t
+let%js create_date : unit -> t = "new Date()"
+let date = create_date ()
+```
+
+## `mel.scope`
+
+Before:
+
+```ocaml
+external node_env : string = "NODE_ENV" [@@mel.scope "process", "env"]
+```
+
+Proposed:
+
+```ocaml
+let%js node_env : string = "process.env.NODE_ENV"
+```
+
+## `mel.uncurry` and `u`
+
+Uncurrying in Melange has historically been complicated, but there might be ways
+to make it simpler to understand and use.
+
+The need for uncurrying OCaml functions in Melange arises when both these
+situations happen:
+- A function written in OCaml is being passed to some JavaScript library through
+  the bindings
+- *And* the function can potentially be called from the JavaScript library
+  itself
+
+Sometimes only the first situation happens. For example, if we write bindings to
+`React.useCallback` we do not need to care about uncurrying, because the `react`
```

Comment:
this is simply incorrect. You can pass the result of `React.useCallback` to a JavaScript library that calls the function.

Reply:
If you pass the result of `React.useCallback` to a JavaScript library that calls the function, there will be some binding to that function. There should be `mel.uncurry` at the boundary layer of _that_ function, not at the `React.useCallback` definition.

If we apply your point more generically, every function in a Melange program should be uncurried because potentially it can be passed to a JavaScript library as callback.